airbyte-integrations/connectors/source-postgres/src/main/java/io/airbyte/integrations/source/postgres/xmin/PostgresXminHandler.java

```diff
@@ -0,0 +1,195 @@
+      final String fullTableName = getFullyQualifiedTableNameWithQuoting(schemaName, tableName,
+          quoteString);
+
+      final String wrappedColumnNames = RelationalDbQueryUtils.enquoteIdentifierList(columnNames, quoteString);
+      final String sql = String.format("SELECT %s FROM %s WHERE xmin::text::bigint > ?",
+          wrappedColumnNames, fullTableName);
+
+      final PreparedStatement preparedStatement = connection.prepareStatement(sql.toString());
+
+      final XminStatus currentStreamXminStatus = xminStateManager.getXminStatus(airbyteStream);
+      if (currentStreamXminStatus != null) {
+        preparedStatement.setLong(1, currentStreamXminStatus.getXminXidValue());
+      } else {
+        preparedStatement.setLong(1, 0L);
+      }
+      LOGGER.info("Executing query for table {}: {}", tableName, preparedStatement);
+      return preparedStatement;
+    } catch (final SQLException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  // Transforms the given iterator to create an {@link AirbyteRecordMessage}
+  private static AutoCloseableIterator<AirbyteMessage> getRecordIterator(
+                                                                          final AutoCloseableIterator<JsonNode> recordIterator,
+                                                                          final String streamName,
+                                                                          final String namespace,
+                                                                          final long emittedAt) {
+    return AutoCloseableIterators.transform(recordIterator, r -> new AirbyteMessage()
+        .withType(Type.RECORD)
+        .withRecord(new AirbyteRecordMessage()
+            .withStream(streamName)
+            .withNamespace(namespace)
+            .withEmittedAt(emittedAt)
+            .withData(r)));
+  }
+
+  // Augments the given iterator with record count logs.
+  private AutoCloseableIterator<AirbyteMessage> augmentWithLogs(final AutoCloseableIterator<AirbyteMessage> iterator,
+      final AirbyteStreamNameNamespacePair pair, final String streamName) {
+    final AtomicLong recordCount = new AtomicLong();
+    return AutoCloseableIterators.transform(iterator,
+        AirbyteStreamUtils.convertFromNameAndNamespace(pair.getName(), pair.getNamespace()),
+        r -> {
+          final long count = recordCount.incrementAndGet();
+          if (count % 10000 == 0) {
+            LOGGER.info("Reading stream {}. Records read: {}", streamName, count);
+          }
+          return r;
+        });
+  }
+
+  private AutoCloseableIterator<AirbyteMessage> augmentWithState(final AutoCloseableIterator<AirbyteMessage> recordIterator,
+      final AirbyteStreamNameNamespacePair pair) {
+    return  AutoCloseableIterators.transform(
+        autoCloseableIterator -> new XminStateIterator(
+            autoCloseableIterator,
+            pair,
+            xminStatus),
```

Comment:
Is `xminStatus` ever changing?
I think this would just put the last saved `xminStatus` back into a state message, no?

Reply:
`xminStatus` is being calculated at the beginning of each sync in `logPresyncDebugData`. So, assuming the stream successfully synced this would indeed emit a new state message (different from the last saved `xminStatus`)

Reply:
OK I see
My concern was with mid sync checkpoint but I see from another response above that we only emit a state by the end because there's no guaranteed order.